src/System.Net.Primitives/src/System/Net/IPAddressPal.Unix.cs

```diff
@@ -0,0 +1,201 @@
+#endif
+            int err = Interop.libc.getnameinfo((Interop.libc.sockaddr*)&sockaddr, (uint)sizeof(Interop.libc.sockaddr_in), buffer, (uint)buffer.Capacity, null, 0, Interop.libc.NI_NUMERICHOST);
+            return unchecked((uint)err);
+        }
+
+        public static unsafe uint Ipv6AddressToString(byte[] address, uint scopeId, StringBuilder buffer)
+        {
+            Debug.Assert(address != null);
+            Debug.Assert(address.Length == IPAddressParser.IPv6AddressBytes);
+            Debug.Assert(buffer != null);
+            Debug.Assert(buffer.Capacity >= IPAddressParser.INET6_ADDRSTRLEN);
+
+            var sockaddr = new Interop.libc.sockaddr_in6 {
+                sin6_family = Interop.libc.AF_INET6,
+                sin6_port = 0,
+                sin6_scope_id = scopeId
+            };
+
+            Debug.Assert(sizeof(Interop.libc.in6_addr) == IPAddressParser.IPv6AddressBytes);
+            for (int i = 0; i < IPAddressParser.IPv6AddressBytes; i++)
+            {
+                sockaddr.sin6_addr.s6_addr[i] = address[i];
+            }
+
+            int err = Interop.libc.getnameinfo((Interop.libc.sockaddr*)&sockaddr, (uint)sizeof(Interop.libc.sockaddr_in6), buffer, (uint)buffer.Capacity, null, 0, Interop.libc.NI_NUMERICHOST);
+            return unchecked((uint)err);
+        }
+
+        public static unsafe uint Ipv4StringToAddress(string ipString, byte[] bytes, out ushort port)
+        {
+            Debug.Assert(ipString != null);
+            Debug.Assert(bytes != null);
+            Debug.Assert(bytes.Length == IPAddressParser.IPv4AddressBytes);
+
+			port = 0;
+
+            var hints = new Interop.libc.addrinfo {
+                ai_flags = Interop.libc.AI_NUMERICHOST | Interop.libc.AI_NUMERICSERV,
+                ai_family = Interop.libc.AF_INET,
+                ai_socktype = 0,
+                ai_protocol = 0
+            };
+
+            Interop.libc.addrinfo* addrinfo = null;
+            int err = Interop.libc.getaddrinfo(ipString, null, &hints, &addrinfo);
+            if (err != 0)
+            {
+                Debug.Assert(addrinfo == null);
+                return unchecked((uint)err);
+            }
+
+            Debug.Assert(addrinfo != null);
+            Debug.Assert(addrinfo->ai_addr != null);
+
+            Interop.libc.sockaddr_in* sockaddr = (Interop.libc.sockaddr_in*)addrinfo->ai_addr;
+
+            uint address = sockaddr->sin_addr.s_addr;
+#if BIGENDIAN
+            bytes[0] = (byte)(address >> 24);
+            bytes[1] = (byte)(address >> 16);
+            bytes[2] = (byte)(address >> 8);
+            bytes[3] = (byte)address;
+#else
+            bytes[0] = (byte)address;
+            bytes[1] = (byte)(address >> 8);
+            bytes[2] = (byte)(address >> 16);
+            bytes[3] = (byte)(address >> 24);
+#endif
+
+            port = sockaddr->sin_port;
+
+            Interop.libc.freeaddrinfo(addrinfo);
+            return unchecked((uint)err);
```

Comment:
@pgavlin For code clarity I would rather have an assert that err is 0 and return 0.